File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    
    var index = 0
    var map = [Int: Int]()
    
    func buildBinaryTree(_ preorder: [Int], _ inorder: [Int], _ root: TreeNode?, _ left: Int, _ right: Int) -> TreeNode?{
    
        if ( left > right){
            return nil
        }
        
        var node = TreeNode(preorder[index])
        var rootIndex = map[preorder[index], default: 0]
        index += 1
        node.left = buildBinaryTree(preorder, inorder, node.left, left ,rootIndex - 1)
        node.right = buildBinaryTree(preorder, inorder, node.right, rootIndex + 1, right)
        
        return node
        
        
    } 
    
    func buildTree(_ preorder: [Int], _ inorder: [Int]) -> TreeNode? {
        
        for index in 0..<inorder.count{
            map[inorder[index], default: 0] = index
        }
        
        return buildBinaryTree(preorder, inorder, nil, 0, inorder.count - 1)
        
    }
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
     unordered_map<int,int> mp ;
    
    TreeNode* constructTree(int *ind , int l , int r,vector<int>& inorder, vector<int>& preorder){
        
        if( l > r){
            return NULL;
        }
        
        TreeNode *newNode = new TreeNode(preorder[*ind]);
        int val = preorder[*ind] ;
        int in = mp[val];
        (*ind)++;
        newNode->left = constructTree( ind ,l , in-1,inorder,preorder);
        newNode->right = constructTree(  ind, in +1 , r ,inorder,preorder);
        return newNode;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int inLength = (int)inorder.size();
        if(inLength == 0 ) return NULL;
        for(int i = 0 ; i < inLength ; i ++ ){
            mp[inorder[i]] = i ;
        }
        int ind = 0 ;
        return constructTree( &ind,0,inLength-1,inorder,preorder);
    }
};